plone/transformchain/zpublisher.py

```diff
@@ -92,11 +92,18 @@ def applyTransformOnSuccess(event):
         response.setBody(transformed)
     # setBody() can deal with byte and unicode strings (and will encode as
     # necessary)...
-    elif isinstance(transformed, six.string_types):
+    elif isinstance(transformed, six.string_types)\
+            or isinstance(transformed, six.binary_type):
         response.setBody(transformed)
     # ... but not with iterables
     else:
-        response.setBody(b''.join(transformed))
+        transformed = map(
+            lambda it: it.decode('utf-8')
+            if isinstance(it, six.binary_type)
+            else it,
+            transformed
+        )
+        response.setBody(''.join(transformed))
```

Comment:
setBody really wants bytes, so I think it would make more sense to encode the items that are text and return bytes here, instead of decoding the items that are binary.